File: steit-csharp-v2/src/Collections/Vector.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

using Steit.Codec;
using Steit.State;

namespace Steit.Collections {
    public sealed class Vector<T> : ReadOnlyCollection<T>, IState {
        public Path Path { get; }

        public Vector(Path? path = null, IList<T>? items = null) : base(items ?? new List<T>()) {
            Typing.CheckPrimitiveOrStateType(typeof(T));
            this.Path = path ?? Path.Root;
        }

        public static Vector<T> Deserialize(IReader reader, Path? path = null) {
            path ??= Path.Root;

            var items = new List<T>();
            var tag = 0U;

            while (!reader.EndOfStream()) {
                // Though Vector doesn't support nested states, passing tags to its children is still helpful.
                items.Add(reader.ReadValue<T>(path, tag++));
            }

            return new Vector<T>(path, items);
        }

        public WireType? GetWireType(UInt32 tag) { return null; }
        public IState? GetNested(UInt32 tag) { return null; }

        public void ReplaceAt(UInt32 tag, WireType wireType, IReader reader, bool shouldNotify) {
            throw new NotSupportedException();
        }

        public bool IsList() { return false; }
        public void ReplayListPush(IReader reader) { throw new NotSupportedException(); }
        public void ReplayListPop() { throw new NotSupportedException(); }

        public bool IsMap() { return false; }
        public void ReplayMapInsert(IReader reader) { throw new NotSupportedException(); }
        public void ReplayMapRemove(IReader reader) { throw new NotSupportedException(); }
    }
}

File: steit-go/pkg/collections/vector.go
package collections

import (
	"github.com/axieinfinity/steit-go/pkg/codec"
	"github.com/axieinfinity/steit-go/pkg/path"
	pathpkg "github.com/axieinfinity/steit-go/pkg/path"
	readerpkg "github.com/axieinfinity/steit-go/pkg/reader"
	statepkg "github.com/axieinfinity/steit-go/pkg/state"
)

var _ statepkg.IState = (*Vector)(nil)

type Vector struct {
	path  *pathpkg.Path
	items []interface{}
	count int
}

func NewVector(path *pathpkg.Path, items []interface{}) *Vector {
	vector := &Vector{}

	if path != nil {
		vector.path = path
	} else {
		vector.path = pathpkg.Root
	}

	if len(items) > 0 {
		vector.items = items
		vector.count = len(items)
	}

	return vector
}

func (v *Vector) Deserialize(reader readerpkg.IReader, path *pathpkg.Path) *Vector {
	if path == nil {
		path = pathpkg.Root
	}

	var items []interface{}
	tag := uint32(0)

	for !reader.EndOfStream() {
		tag = tag + 1
		items = append(items, readerpkg.ReadValue(reader, path, tag))
	}

	return NewVector(path, items)
}

func (v *Vector) GetPath() *path.Path {
	return v.path
}

func (v *Vector) GetWireType(tag uint32) *codec.WireType {
	return nil
}

func (v *Vector) GetNested(tag uint32) statepkg.IState {
	return nil
}

func (v *Vector) ReplaceAt(tag uint32, wireType codec.WireType, reader readerpkg.IReader, shouldNotify bool) {
	if int(tag) >= v.count {
		panic("index out of range")
	}
	panic("not supported")
}

func (v *Vector) ReplayListPush(reader readerpkg.IReader) {
	panic("not supported")
}

func (v *Vector) ReplayListPop() {
	panic("not supported")
}

func (v *Vector) ReplayMapRemove(key uint32) {
	panic("not supported")
}
